Add SubjectKeyId to certificates

felixfontein opened the pull request:
According to https://tools.ietf.org/html/rfc5280.html#section-4.2.1.2 all certificates including the Basic Constraints extension should have a `SubjectKeyId`. This PR makes sure the certificates generated by Pebble have one: it adds random `SubjectKeyId`s to all certificates. According to https://golang.org/pkg/crypto/x509/#CreateCertificate, the intermediate and leaf certs should get `AuthorityKeyId` automatically added.

felixfontein commented:
There's a potential problem: this causes the leaf certificate to get a `AuthorizedKeyIdentifier` extension, which points to the main intermediate. But this certificate is also delivered as an alternative leaf certificate with an alternative chain, which would then be wrong. If I understand this correctly, there need to be one leaf certificate per alternate chain.

alexzorin commented:
For alternate chains in ACME, it's necessarily the case that `issuer.ski == leaf.aki`, because it's a requirement of RFC5280. Anything else is an invalid configuration.

felixfontein commented:
@alexzorin that's correct, though it might encourage ACME clients to assume that this is always the case with alternate chains returned by ACME servers. I've extended Pebble so that it creates a different leaf certificate per alternate chain, so that each alternate chain is a proper chain according to RFC5280.

alexzorin commented:
I don't think issuing two leaf certificates is what is intended by ACME.

>The server MAY provide one or more link relation header fields
   [RFC8288] with relation "alternate".  Each such field SHOULD express
   an alternative certificate chain **starting with the same end-entity
   certificate**

The SKI is supposed to uniquely and consistently identify the public key of the certificate (https://tools.ietf.org/html/rfc5280#section-4.2.1.2):

>For CA certificates, subject key identifiers SHOULD be derived from
   the public key or a method that generates unique values
 
Pebble already correctly re-uses the same private key for every intermediate that it uses to actually sign certificates:

https://github.com/letsencrypt/pebble/blob/bb4787328a963c67e223f3f37741908aff2a6e54/ca/ca.go#L249-L256

What follows is that every intermediate's SKI is identical (according to RFC5280, but not according to this PR so far).

This leads to a very simple solution: set the SKI of the intermediate according to its public key (you can use [this](https://play.golang.org/p/58cpHs-N7lV)), and everything else falls into place (no SKI/AKI mismatches between the leaf and any of the alternate chains).

felixfontein commented:
> I don't think issuing two leaf certificates is what is intended by ACME.
> 
> > The server MAY provide one or more link relation header fields
> > [RFC8288] with relation "alternate".  Each such field SHOULD express
> > an alternative certificate chain **starting with the same end-entity
> > certificate**

Ah. I thought I'd read that somewhere, but didn't see it when looking again yesterday.

> The SKI is supposed to uniquely and consistently identify the public key of the certificate (https://tools.ietf.org/html/rfc5280#section-4.2.1.2):
> 
> > For CA certificates, subject key identifiers SHOULD be derived from
> > the public key or a method that generates unique values

So a random value is fine as well :) But all intermediates should better have the same one.

> This leads to a very simple solution: set the SKI of the intermediate according to its public key (you can use [this](https://play.golang.org/p/sJ_zLczoP20) to calculate it), and everything else falls into place (no SKI/AKI mismatches between the leaf and any of the alternate chains).

Where does that code come from? (Is it yours?) What license does it have? Can we use it in Pebble?

felixfontein commented:
I've implemented this now. I've exchanged the random SKI with the code provided by @alexzorin in the last commit; in case it turns out we can't use this code, we can simply remove that commit (or replace it by something else).

alexzorin commented:
The SKI code [is taken from cfssl](https://github.com/cloudflare/cfssl/blob/b94e044bb51ec8f5a7232c71b1ed05dbe4da96ce/signer/signer.go#L221-L244), which is what Boulder uses to implement its CA component. @cpu has previously noted that cfssl is too big a dependency to include in Pebble, but hopefully what is effectively a 3-liner can be copied under fair use.

felixfontein commented:
I guess that's up to @jsha and @cpu to decide. I've added error handling to the snippet (similar to the original), now it's a bit more than three lines :) I've also added the source in a comment.

cpu commented:
@felixfontein Can you help me explain what motivated this change? I agree the certificates being issued by Pebble now don't match the SubjectKeyId requirement from RFC 5280 but without understanding what complications that is causing its hard to evaluate the need for the fix. 

It seems like a sensible change but if there is a larger perspective in mind to have Pebble issue certificates that fully adhere to 5280 and the baseline requirements vs a one-off fix there are probably going to be other spec considerations that aren't being adhered to.

cpu commented:
> hopefully what is effectively a 3-liner can be copied under fair use.

Agreed, I don't think there is a sensible alternative way to do what this code is doing with Go without effectively copying the function.

felixfontein commented:
> @felixfontein Can you help me understand what motivated this change? I agree the certificates being issued by Pebble now don't match the SubjectKeyId requirement from RFC 5280 but without understanding what complications that is causing its hard to evaluate the need for the fix.
> 
> It seems like a sensible change but if there is a larger perspective in mind to have Pebble issue certificates that fully adhere to 5280 and the baseline requirements vs a one-off fix there are probably going to be other spec considerations that aren't being adhered to.

For me the main motivation is that I'm implementing alternate certificate chain selection in my client, and selecting a chain by `authorityKeyID` is one of the options I want to offer. As far as I understand it, using `authorityKeyID` (which needs `subjectKeyID` to work) is the best way to do this.

Since Pebble isn't including `subjectKeyID` (and thus also no `authorityKeyID` down the chain), I currently have no way to test this with integration tests.

felixfontein commented:
(Rebasing so it's easier for me to run all tests at once.)

felixfontein commented:
I've rebased to remove the conflict (with #276 merged). During that process I squashed all commits, to avoid having to do the same conflict resolution several times.

felixfontein commented:
@alexzorin @cpu @jsha thanks a lot for your feedback, reviewing and merging! :)
